### DISCUSSION TITLE
baresip, multicast, Raspberry Pi, and Innomaker RPI HiFi AMP HAT

### DISCUSSION THREAD

**Discussion by timmrogers** [baresip/baresip#3022] [Q&A]
I'm having an odd problem with using baresip to play multicast audio with a Raspberry Pi and the Innomaker RPI HiFi AMP HAT (https://www.inno-maker.com/product/hifi-amp-hat/ ).

I have build about 30 baresip PAs using both old Mac Minis and Raspberrry Pis. They are configured as SIP endpoints and to accept multicast audio and work fine. 

My existing Raspberry Pi PAs use the built-in audio (for the Raspberry Pi 4s) or a USB audio interface (for the Raspberry Pi 5s).

My goal now it so build some Raspberry PI 4s with the Innomaker RPI HiFi AMP HAT so I can connect them directly to loudspeakers.

I have a prototype working fine at a SIP endpoint, but, while it receives the multicast audio - it makes no sound. That is, it plays sound from thee ALSA aplay tool and from baresip when I page its extension, but (while I can see that it receives the incoming multicast audio from the console) it makes no sound when it gets multicast audio. 

I did have to configure baresip to use use the dummy interface for the mic and lock it to 48000 hz to get it to play SIP pages.

~/.baresip/config

<pre>
...

# Audio
#audio_path		/usr/local/share/baresip
audio_player		alsa,default
# audio_source		alsa,default
audio_source		aubridge,dummy
audio_alert		alsa,default
ausrc_srate		48000
auplay_srate		48000
# ausrc_srate		44100
# auplay_srate		44100
...
</pre>

I'm wondering if there is something similar that I am missing in order to make the Innomaker amp work with multicast audio.

Thanks everyone. I appreciate any ideas or suggestions you may have!

**Comment by cspiel1** [comment]
We shortly moved multicast module to baresip-apps.
See `baresip-apps/modules/multicast/receiver.c:153`!
There you see that the codec is selected depending on the payload type. Note, that for RTP audio streaming without any signaling (SDP) dynamic payload types are not supported. Thus other codecs than G.711 and G.722 are not supported.

- G.711 with 8 kHz
- G.722 with 16 kHz

Resampling, more concretely up-sampling need to be done. This can be done by module `auresamp` which need to be loaded with
adding to ~/.baresip/config:
```
module   auresamp.so
```

Note, that only divisible input/output sample rates are supported! In this case (16000 -> 48000 , 8000 -> 48000) should work.

**Comment by timmrogers** [comment]
Thanks for the reply.

I appreciate the heads up about the move to baresip-apps. I'll make sure to include the module from there the next time I compile.

I am using G.722 for the audio multicasts, and, as I said, it works nicely on all of my the baresip instances. Only the one running the Innomaker RPI HiFi AMP HAT fails, and only for multicast, not SIP pages. 

The audio is already formatted to 48000 Hz, but I also have the auresamp.so module available and enabled. 

If there some debugging or other information that I could share that might be helpful?

Thanks again!

**Comment by timmrogers** [comment]
So, it looks like baresip is converting incoming SIP pages from 16000 Hz to 48000 Hz, but not incoming multicast audio. 

<pre>
...
multicast receiver: start addr=224.0.36.10:10010 prio=2 enabled=1 state=receiving
alsa: reset: srate=16000, ch=1, num_frames=320, pcmfmt=S16_LE
alsa: playback started (default)
multicast receiver: start addr=224.0.36.10:10010 prio=2 enabled=1 state=running
auresamp: resample decoder 16000/1 --> 16000/1
multicast receiver: EOS addr=224.0.36.10:10010 prio=2 enabled=1 state=running
alsa: stopping playback thread (default)
ua: no call to resume
ua: sipsess connect via UDP 172.16.0.6:5060 --> 172.16.5.141:35312
7341@172.16.0.6: selected for 7341-0x55acf20bb0
ua: using connection-address 172.16.0.6 of SDP offer
call: alloc with params laddr=172.16.5.141, af=AF_INET, use_rtp=1
call: use_video=0
call: answering call on line 1 from sip:7230@172.16.0.6 with 200
audio: peer changed ptime_tx 40ms -> 20ms
stream: update 'audio'
stream: disable audio RTP receiver
stream: disable audio RTP sender
stream: enable audio RTP receiver
stream: enable audio RTP sender
stream: audio: starting RTCP with remote 172.16.0.6:15151
audio: update
audio: Set audio decoder: G722 16000Hz 1ch
audio: start
alsa: reset: srate=48000, ch=1, num_frames=1920, pcmfmt=S16_LE
alsa: playback started (default)
audio: player started with sample format S16LE
audio: Set audio encoder: G722 16000Hz 1ch
audio: start
aubridge: created device 'dummy'
audio: source started with sample format S16LE
audio tx pipeline:    aubridge ---> aubuf ---> auconv ---> auresamp ---> G722
audio rx pipeline:        alsa <--- aubuf <--- auconv <--- auresamp <--- G722
call: stream start (active=1)
audio: Set audio decoder: G722 16000Hz 1ch
audio: start
stream: enable audio RTP receiver
stream: enable audio RTP sender
7341@172.16.0.6: Call established: sip:7230@172.16.0.6
stream: incoming rtp for 'audio' established, receiving from 172.16.0.6:15150
auresamp: resample decoder 16000/1 --> 48000/1
audio_recv: create audio buffer [20 - 160 ms] [1920 - 15360 bytes]
sip:7230@172.16.0.6: session closed: Connection reset by peer [104]
stream: disable audio RTP receiver
...
</pre>

Can someone confirm that is what these logs say? And, if so, any suggestions as to how to ask baresip to convert for multicast audio?

Thank you!

**Reply by cspiel1** [reply]
Yes, the sample rate for the audio player is taken from the codec, which for G.722 is 16kHz.
I think this should be changed in order to respect the setting `auplay_srate`. The codec srate should be used by default. The config may change this.

@cHuberCoffee do you agree?

**Reply by cspiel1** [reply]
I think it is a bug. Let's proceed here: https://github.com/baresip/baresip/issues/3024
